internal/pkg/agent/application/upgrade/artifact/download/http/verify_backoff_rtt.go

```diff
@@ -0,0 +1,58 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package http
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/errors"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+)
+
+func WithBackoff(rtt http.RoundTripper, logger *logger.Logger) http.RoundTripper {
+	if rtt == nil {
+		rtt = http.DefaultTransport
+	}
+
+	return &BackoffRoundTripper{next: rtt, logger: logger}
+}
+
+type BackoffRoundTripper struct {
+	next   http.RoundTripper
+	logger *logger.Logger
+}
+
+func (btr *BackoffRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	exp := backoff.NewExponentialBackOff()
+	boCtx := backoff.WithContext(exp, req.Context())
+
+	opNotify := func(err error, retryAfter time.Duration) {
+		btr.logger.Warnf("request failed: %s, retrying in %s", err, retryAfter)
+	}
+
+	var resp *http.Response
+	var err error
+	opFunc := func() error {
+		resp, err = btr.next.RoundTrip(req)
+		if err != nil {
+			return err
+		}
```

Comment:
This lint error should be fixed. Should probably call close no matter what.

Reply:
Looked into this with @belimawr . The linter is most probably complaining here because the response is used in a closure and not passed onto anything else. Closing the response body here causes an error when the caller tries to read from it. The response is closed after each unsuccessful attempt so that we don't leak resources. When an attempt is successful, the response is sent back up the call stack. This is similar to how it is implemented in [FleetAuthRoundTripper](https://github.com/elastic/elastic-agent/blob/d9ee653f10ccb8fc30828ddb9c33dd94280c53e9/internal/pkg/fleetapi/client/round_trippers.go#L44). Another option was to read the whole response into a buffer, close the body, and then assign the bufffer back to the response body after wrapping it with `io.NopCloser`. This is how the [DebugRoundTripper](https://github.com/elastic/elastic-agent/blob/d9ee653f10ccb8fc30828ddb9c33dd94280c53e9/internal/pkg/remote/round_trippers.go#L53) handles the request body. While it is a valid solution, it may cause issues in case there are other midleware in the rountripper chain where there may be a response body size limiter, and it also doesn't seem to be necessary here as the current implementation works as expected. To resolve the linting problem I added a `nolint` comment.

Reply:
Thanks for the detailed explanation. Glad to see it was also added to the code.